fv3core/stencils/a2b_ord4.py

```diff
@@ -22,485 +36,435 @@
+        # {
+
+        with horizontal(
+            region[i_start - 1 : i_start + 1, :], region[i_end : i_end + 2, :]
+        ):
+            q2 = (qin[-1, 0, 0] * dxa + qin * dxa[-1, 0]) / (dxa[-1, 0] + dxa)
+        with horizontal(region[i_start, j_start + 1 : j_end + 1]):
+            qout = qout_x_edge(qin, dxa, edge_w, q2)
+        with horizontal(region[i_end + 1, j_start + 1 : j_end + 1]):
+            qout = qout_x_edge(qin, dxa, edge_e, q2)
+
+        with horizontal(
+            region[:, j_start - 1 : j_start + 1], region[:, j_end : j_end + 2]
+        ):
+            q1 = (qin[0, -1, 0] * dya + qin * dya[0, -1]) / (dya[0, -1] + dya)
+        with horizontal(region[i_start + 1 : i_end + 1, j_start]):
+            qout = qout_y_edge(qin, dya, edge_s, q1)
+        with horizontal(region[i_start + 1 : i_end + 1, j_end + 1]):
+            qout = qout_y_edge(qin, dya, edge_n, q1)
+
+        # compute_qx
+        qx = ppm_volume_mean_x(qin)
+        with horizontal(region[i_start, :]):
+            qx = qx_edge_west(qin, dxa)
+        with horizontal(region[i_start + 1, :]):
+            qx = qx_edge_west2(qin, dxa, qx)
+
+        with horizontal(region[i_end + 1, :]):
+            qx = qx_edge_east(qin, dxa)
+        with horizontal(region[i_end, :]):
+            qx = qx_edge_east2(qin, dxa, qx)
+
+        # compute_qy
+        qy = ppm_volume_mean_y(qin)
+        with horizontal(region[:, j_start]):
+            qy = qy_edge_south(qin, dya)
+        with horizontal(region[:, j_start + 1]):
+            qy = qy_edge_south2(qin, dya, qy)
+
+        with horizontal(region[:, j_end + 1]):
+            qy = qy_edge_north(qin, dya)
+        with horizontal(region[:, j_end]):
+            qy = qy_edge_north2(qin, dya, qy)
+
+        # compute_qxx
+        qxx = lagrange_y(qx)
+        with horizontal(region[:, j_start + 1]):
+            qxx = cubic_interpolation_south(qx, qout, qxx)
+        with horizontal(region[:, j_end]):
+            qxx = cubic_interpolation_north(qx, qout, qxx)
+
+        # compute_qyy
+        qyy = lagrange_x(qy)
+        with horizontal(region[i_start + 1, :]):
+            qyy = cubic_interpolation_west(qy, qout, qyy)
+        with horizontal(region[i_end, :]):
+            qyy = cubic_interpolation_east(qy, qout, qyy)
+
+        with horizontal(region[i_start + 1 : i_end + 1, j_start + 1 : j_end + 1]):
+            qout = 0.5 * (qxx + qyy)
+        # }
+
+        if __INLINED(REPLACE):
+            qin = qout
+
+
+def _make_grid_storage_2d(grid_array: gt4py.storage.Storage, index: int = 0):
+    grid = spec.grid
+    return gt4py.storage.from_array(
+        grid_array[:, :, index],
+        backend=global_config.get_backend(),
+        default_origin=grid.compute_origin()[:-1],
+        shape=grid_array[:, :, index].shape,
+        dtype=grid_array.dtype,
+        mask=(True, True, False),
     )
 
 
-def compute(qin, qout, kstart=0, nk=None, replace=False):
+def compute(
```

Comment:
Would be nice to start a docstring on this one (at least with just a TODOC) since it gets called in so many different ways.

Reply:
Good idea. I'll write one!